### PR TITLE
Add email template set to Share, DDSDelivery and S3Delivery

### PR DESCRIPTION
**Pull request by johnbradley**
WIP

### CODE REVIEW

**Comment by johnbradley** [comment]
Right now the new `email_template_set` field allows null. Ideally this would be filled in by a migration.
There is no direct link from a DDSDelivery to the User that created it and thereby the associated email template. However it looks like our history plugin records `history_user` and `history_type` records. https://github.com/Duke-GCB/D4S2/blob/f4a26705d349c0887258e7ceb04a53e0852f0986/d4s2_api/models.py#L139

It may be possible to find the history_type of created for each share and delivery then find the email template set based on UserEmailTemplateSet.

**Comment by johnbradley** [comment]
Need to add tests for when email_template_set is null.

File: d4s2_api_v2/api.py [hunk 1]
```diff
@@ -10,12 +10,14 @@
 from d4s2_api_v2.serializers import DDSUserSerializer, DDSProjectSerializer, DDSProjectTransferSerializer, \
     UserSerializer, S3EndpointSerializer, S3UserSerializer, S3BucketSerializer, S3DeliverySerializer, \
     DDSProjectPermissionSerializer, DDSDeliveryPreviewSerializer
-from d4s2_api.models import DDSDelivery, S3Endpoint, S3User, S3UserTypes, S3Bucket, S3Delivery, UserEmailTemplateSet
+from d4s2_api.models import DDSDelivery, S3Endpoint, S3User, S3UserTypes, S3Bucket, S3Delivery
 from d4s2_api_v1.api import AlreadyNotifiedException, get_force_param, build_accept_url, DeliveryViewSet, \
-    EMAIL_TEMPLATES_NOT_SETUP_MSG
+    get_email_template_for_request, populate_email_template_in_request, prevent_email_template_set_in_request, \
```
Comment (dleehr) [inline d4s2_api_v2/api.py]: Importing this many functions from v1 for this change seems like a code smell. As noted earlier, I think some of it can be handled by the serializer, but the rest of the logic is probably better encapsulated in a Mixin.

File: d4s2_api_v1/api.py [hunk 2]
```diff
@@ -22,6 +24,42 @@ def build_accept_url(request, transfer_id, delivery_type):
     return accept_url
 
 
+def prevent_email_template_set_in_request(request):
```
[the remark this answers is not in the document]
Comment (johnbradley) [reply d4s2_api_v1/api.py]: The serializer needs to fill in the `email_template_set` with the value passed by the API layer when the object being serialized is created.

[the remark this answers is not in the document]
Comment (dleehr) [reply d4s2_api_v1/api.py]: Not sure I follow. Can you walk me through it? If a GitHub comment isn't a good place to explain, I can do a quick call.

[the remark this answers is not in the document]
Comment (dleehr) [reply d4s2_api_v1/api.py]: One snippet that may help

https://github.com/Duke-GCB/D4S2/blob/f4a26705d349c0887258e7ceb04a53e0852f0986/d4s2_api_v2/api.py#L270-L290

[the remark this answers is not in the document]
Comment (johnbradley) [reply d4s2_api_v1/api.py]: After talking this through I get this now. Making the field `read_only` in the serializer only makes it read only from the request payload. The api layer can reach into the serializer and fill in values on the object that the serializer has marked as `read_only`.

File: ownership/test_views.py [hunk 3]
```diff
@@ -207,6 +210,67 @@ def test_get_not_allowed(self):
         response = self.client.get(url)
         self.assertEqual(response.status_code, status.HTTP_405_METHOD_NOT_ALLOWED)
 
+    @patch('switchboard.dds_util.DDSUtil')
+    @patch('switchboard.dds_util.DDSProjectTransfer')
+    @patch('d4s2_api.utils.Message')
+    def test_receiving_user_can_accept_deliveries_without_email_template_set(self, mock_message,
```
Comment (johnbradley) [inline ownership/test_views.py]: High level root issue test `test_receiving_user_can_accept_deliveries_without_email_template_set`.

File: ownership/test_views.py [hunk 4]
```diff
@@ -207,6 +210,67 @@ def test_get_not_allowed(self):
...
+        response = self.client.post(url, {'transfer_id': transfer_id}, follow=True)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+        delivery.refresh_from_db()
+        self.assertEqual(delivery.state, State.ACCEPTED)
+
+    @patch('switchboard.dds_util.DDSUtil')
+    @patch('switchboard.dds_util.DDSProjectTransfer')
+    @patch('d4s2_api.utils.Message')
+    def test_delivery_emails_always_use_senders_email_template_set(self, mock_message, mock_dds_project_transfer,
```
Comment (johnbradley) [inline ownership/test_views.py]: High level root issue test `test_delivery_emails_always_use_senders_email_template_set `.

File: d4s2_api_v1/api.py [hunk 5]
```diff
@@ -24,48 +25,44 @@ def build_accept_url(request, transfer_id, delivery_type):
...
+class ModelWithEmailTemplateSetMixin(object):
+    def create(self, request, *args, **kwargs):
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        serializer.save(email_template_set=self.get_email_template_for_request(request))
+        headers = self.get_success_headers(serializer.data)
+        return Response(serializer.data, status=status.HTTP_201_CREATED, headers=headers)
+
+    @staticmethod
+    def get_email_template_for_request(request):
```
Comment (dleehr) [inline d4s2_api_v1/api.py]: Can you elaborate on why this is a static method? It seems like instance-level functionality. The instance will have a `self.request` property, so if it were an instance method, `request` doesn't need to be passed as an argument.

Unfortunately, I can't seem to find documentation about where exactly `self.request` gets assigned, but it seems to be pretty fundamental in django class-based views

¯\\_(ツ)_/¯

File: d4s2_api_v1/api.py [hunk 6]
```diff
@@ -24,48 +25,44 @@ def build_accept_url(request, transfer_id, delivery_type):
...
+        :return: EmailTemplateSet
+        """
+        try:
+            user_email_template_set = UserEmailTemplateSet.objects.get(user=request.user)
+            return user_email_template_set.email_template_set
+        except UserEmailTemplateSet.DoesNotExist:
+            raise ValidationError(EMAIL_TEMPLATES_NOT_SETUP_MSG)
+
+    @staticmethod
+    def prevent_null_email_template_set(obj):
```
Comment (dleehr) [inline d4s2_api_v1/api.py]: Can you explain why this method is static too? `self.get_object()`will return the current model object.

File: ownership/test_views.py (hunk 3 above)
Comment (dleehr) [reply ownership/test_views.py]: 👍

File: ownership/test_views.py (hunk 4 above)
Comment (dleehr) [reply ownership/test_views.py]: The case I had in mind was to show when a sender creates a delivery via the API, their template set is used for all emails related to the delivery, regardless of other template sets in the database or if the recipient user has a template set or not. Really more of an end-to-end test for a scenario.

I see that the API lookup logic is unit tested, but thought this specific issue was a good place to have a higher level test.

File: d4s2_api_v1/api.py (hunk 6 above)
Comment (johnbradley) [reply d4s2_api_v1/api.py]: As far as I can tell `self.get_object()` tries to query a new object from the database.
Having two instances of our object floating around seemed problematic.

File: d4s2_api_v1/api.py (hunk 5 above)
Comment (johnbradley) [reply d4s2_api_v1/api.py]: I wanted to make the code really obvious what was going on and minimize surprise. When you see a method that receives a `requests` parameter it is rather surprising that `self.requests` is also the same value.  I was able to find a couple places in DRF where they use `self.request` but nothing that spells it out. It's definitely there in the code. I will try out switching to the instance version and we can see how that looks.

**Comment by johnbradley** [comment]
> show when a sender creates a delivery via the API, their template set is used for all emails related to the delivery, regardless of other template sets in the database or if the recipient user has a template set or not. Really more of an end-to-end test for a scenario

Where would this test live? It would have to test parts of `d4s2_api_v1` and `ownership`.

File: d4s2_api_v1/api.py (hunk 5 above)
Comment (johnbradley) [reply d4s2_api_v1/api.py]: It was a `@staticmethod` so it could be used in the DeliveryPreview. The DeliveryPreview api endpoint doesn't make use of the create method so mixing an object that adds a create method seemed odd just to use one of the methods. You can see how this looks as an instance method in the latest commit:
```
class DeliveryPreviewView(ModelWithEmailTemplateSetMixin, generics.CreateAPIView):
...
    def create(self, request, *args, **kwargs):
        email_template_set = self.get_email_template_for_request()

        serializer = self.get_serializer(data=request.data)
        serializer.is_valid(raise_exception=True)
...
```

**Comment by dleehr** [comment]
> Where would this test live? It would have to test parts of `d4s2_api_v1` and `ownership`.

That's a good question. This would be an integration test (or acceptance if we could loop in the external pieces that people use) so as you point out, it rises above a single app. I don't know that Django has an obvious convention about those (or where they would go)

We could create a subdirectory in the repo called `integration_tests` and put the test file there. This can be a simple directory - doesn't need to be a django "app" (so no models/views/migrations, not registered in settings.py).

We do a similar thing in bespin-api. There's a bespin-flavored OAuth backend in [bespin-api/bespin/backend.py](https://github.com/Duke-GCB/bespin-api/blob/master/bespin/backend.py). That directory isn't a django "app", but the backend is used by [settings_base.py](https://github.com/Duke-GCB/bespin-api/blob/master/bespin/settings_base.py#L113). The [backend_test.py](https://github.com/Duke-GCB/bespin-api/blob/master/bespin/test_backend.py) is in the same directory and gets picked up by the test runner.

**Comment by johnbradley** [comment]
@dleehr This is ready for another look. 
The current version uses instance methods for the Mixin.
The only part that is still bugging me is how this works with `DeliveryPreviewView`:
```
class DeliveryPreviewView(ModelWithEmailTemplateSetMixin, generics.CreateAPIView):
    permission_classes = (permissions.IsAuthenticated,)
    serializer_class = DDSDeliveryPreviewSerializer


    def create(self, request, *args, **kwargs):
        email_template_set = self.get_email_template_for_request()
...
```
We mixin another `create` method but we don't actually use it. We just do so to let the actual `create` method use the `get_email_template_for_request` method.

File: integration_tests/test_integration.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        if user_id == self.recipient_dds_id:
+            return Mock(full_name='Recipient', email=self.recipient_user.email)
+        raise ValueError("Invalid user id:" + user_id)
+
+    @patch('d4s2_api_v1.api.DDSUtil')
+    @patch('switchboard.dds_util.DDSUser')
+    @patch('switchboard.dds_util.DDSProjectTransfer')
+    @patch('switchboard.dds_util.RemoteStore')
+    @patch('d4s2_api.utils.Message')
+    def test_accepted_delivery_uses_senders_email_template_set(self, mock_message, mock_remote_store,
```
Comment (johnbradley) [inline integration_tests/test_integration.py]: Here is the integration test where a delivery sender creates and sends a delivery and then a recipient accepts it.

**Comment by dleehr** [comment]
> We mixin another `create` method but we don't actually use it. We just do so to let the actual `create` method use the `get_email_template_for_request` method.

I think that's fine. There might be one more layer of refactoring to get that pebble out of the shoe, but I don't want to drag this on any further.
